src/utils/retry.py

```
@@ -0,0 +1,377 @@
+            if attempt < config.max_attempts - 1:
+                delay = calculate_delay(attempt, config)
+                logger.warning(
+                    f"Attempt {attempt + 1}/{config.max_attempts} failed "
+                    f"({type(e).__name__}). Retrying in {delay:.2f}s: {e}"
+                )
+                time.sleep(delay)
+            else:
+                logger.error(
+                    f"All {config.max_attempts} attempts failed. "
+                    f"Last error: {type(e).__name__}: {e}"
+                )
```

Ruff (0.14.3)

Use `logging.exception` instead of `logging.error`

Replace with `exception`

(TRY400)